File: ruby/lib/scl/lexer.rb
require_relative "errors"

module SCL
  Token = Struct.new(:type, :value, :line, :column)

  module TokenType
    IDENTIFIER = :identifier
    DOUBLE_COLON = :double_colon
    BOOL = :bool
    STR = :str
    NUM = :num
    FL = :fl
    ML = :ml
    CLASS = :class
    LIST = :list
    DYNAMIC = :dynamic
    LBRACE = :lbrace
    RBRACE = :rbrace
    LPAREN = :lparen
    RPAREN = :rparen
    COMMA = :comma
    STRING = :string
    MULTILINE_STRING = :multiline_string
    NUMBER = :number
    FLOAT = :float
    BOOLEAN = :boolean
    COMMENT = :comment
    NEWLINE = :newline
    EOF = :eof
  end

  class Lexer
    def initialize(text)
      @text = text
      @pos = 0
      @line = 1
      @col = 1
      @tokens = []
    end

    def self.tokenize(text)
      new(text).tokenize
    end

    def tokenize
      while @pos < @text.length
        skip_whitespace
        break if peek.nil?

        case peek
        when "["
          @tokens << read_comment
        when "\n"
          add(TokenType::NEWLINE, "\n")
          advance
        when ":"
          if peek(1) == ":"
            add(TokenType::DOUBLE_COLON, "::")
            advance(2)
          else
            error("Unexpected ':'")
          end
        when "{", "}", "(", ")", ","
          map = { "{" => TokenType::LBRACE, "}" => TokenType::RBRACE, "(" => TokenType::LPAREN, ")" => TokenType::RPAREN, "," => TokenType::COMMA }
          add(map[peek], peek)
          advance
        when "\""
          @tokens << read_string
        when "'"
          @tokens << read_multiline_string
        when "-"
          if digit?(peek(1))
            @tokens << read_number
          else
            error("Expected digit after '-'")
          end
        else
          if digit?(peek)
            @tokens << read_number_or_identifier
          elsif alpha_or_underscore?(peek)
            @tokens << read_identifier
          else
            error("Unexpected character: #{peek}")
          end
        end
      end
      @tokens << Token.new(TokenType::EOF, nil, @line, @col)
      @tokens
    end

    private

    def add(type, value)
      @tokens << Token.new(type, value, @line, @col)
    end

    def peek(offset = 0)
      i = @pos + offset
      return nil if i >= @text.length
      @text[i]
    end

    def advance(n = 1)
      n.times do
        ch = @text[@pos]
        @pos += 1
        if ch == "\n"
          @line += 1
          @col = 1
        else
          @col += 1
        end
      end
    end

    def skip_whitespace
      while peek == " " || peek == "\t"
        advance
      end
    end

    def read_comment
      line, col = @line, @col
      advance # skip '['
      buf = +""
      while peek && peek != "]"
        buf << peek
        advance
      end
      error("Unclosed comment") unless peek == "]"
      advance # skip ']'
      Token.new(TokenType::COMMENT, buf.strip, line, col)
    end

    def read_string
      line, col = @line, @col
      advance # skip "
      buf = +""
      while peek && peek != "\""
        if peek == "\\"
          advance
          case peek
          when "n" then buf << "\n"
          when "t" then buf << "\t"
          when "\"", "\\" then buf << peek
          else buf << peek
          end
          advance
        else
          buf << peek
          advance
        end
      end
      error("Unclosed string") unless peek == "\""
      advance # skip "
      Token.new(TokenType::STRING, buf, line, col)
    end

    def read_multiline_string
      line, col = @line, @col
      advance # skip '
      buf = +""
      while peek && peek != "'"
        buf << peek
        advance
      end
      error("Unclosed multiline string") unless peek == "'"
      advance # skip '
      Token.new(TokenType::MULTILINE_STRING, buf, line, col)
    end

    def read_number
      line, col = @line, @col
      buf = +""
      if peek == "-"
        buf << peek
        advance
        error("Expected digit after '-'") unless digit?(peek)
      end
      has_dot = false
      while peek && (digit?(peek) || peek == ".")
        if peek == "."
          break if has_dot
          has_dot = true
        end
        buf << peek
        advance
      end
      if has_dot
        Token.new(TokenType::FLOAT, buf.to_f, line, col)
      else
        Token.new(TokenType::NUMBER, buf.to_i, line, col)
      end
    end

    def read_number_or_identifier
      # lookahead digits then alpha/_ -> identifier
      line, col = @line, @col
      p = @pos
      while p < @text.length && digit?(@text[p])
        p += 1
      end
      if p < @text.length && (alpha?(@text[p]) || @text[p] == "_")
        read_identifier
      else
        read_number
      end
    end

    def read_identifier
      line, col = @line, @col
      buf = +""
      while peek && (alpha?((c = peek)) || digit?(c) || c == "_" || c == "-")
        buf << c
        advance
      end
      keywords = {
        "bool" => TokenType::BOOL,
        "str" => TokenType::STR,
        "num" => TokenType::NUM,
        "fl" => TokenType::FL,
        "ml" => TokenType::ML,
        "class" => TokenType::CLASS,
        "list" => TokenType::LIST,
        "dynamic" => TokenType::DYNAMIC
      }
      booleans = {
        "true" => true, "false" => false, "yes" => true, "no" => false
      }
      if keywords.key?(buf)
        Token.new(keywords[buf], buf, line, col)
      elsif booleans.key?(buf)
        Token.new(TokenType::BOOLEAN, booleans[buf], line, col)
      else
        Token.new(TokenType::IDENTIFIER, buf, line, col)
      end
    end

    def digit?(c) = c && c >= "0" && c <= "9"
    def alpha?(c) = c && ((c >= "A" && c <= "Z") || (c >= "a" && c <= "z"))
    def alpha_or_underscore?(c) = alpha?(c) || c == "_"

    def error(msg)
      raise SyntaxError.new(msg, line: @line, column: @col)
    end
  end
end

File: python/scl_parser/lexer.py
from typing import Any, List, Optional

from .errors import SCLSyntaxError
from .tokens import Token, TokenType


class Lexer:
    def __init__(self, text: str):
        self.text = text
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []
    
    def error(self, msg: str):
        raise SCLSyntaxError(msg, self.line, self.column)
    
    def peek(self, offset: int = 0) -> Optional[str]:
        pos = self.pos + offset
        if pos < len(self.text):
            return self.text[pos]
        return None
    
    def advance(self) -> Optional[str]:
        if self.pos < len(self.text):
            char = self.text[self.pos]
            self.pos += 1
            if char == '\n':
                self.line += 1
                self.column = 1
            else:
                self.column += 1
            return char
        return None
    
    def skip_whitespace(self):
        while self.peek() in ' \t':
            self.advance()
    
    def read_comment(self) -> Token:
        start_line = self.line
        start_col = self.column
        self.advance()
        comment = ""
        while self.peek() and self.peek() != ']':
            comment += self.advance()
        if self.peek() != ']':
            self.error("Unclosed comment")
        self.advance()
        return Token(TokenType.COMMENT, comment.strip(), start_line, start_col)
    
    def read_string(self) -> Token:
        start_line = self.line
        start_col = self.column
        self.advance()
        string = ""
        while self.peek() and self.peek() != '"':
            if self.peek() == '\\':
                self.advance()
                next_char = self.peek()
                if next_char is None:
                    self.error("Unexpected end of string after backslash")
                if next_char == 'n':
                    string += '\n'
                elif next_char == 't':
                    string += '\t'
                elif next_char in ('"', '\\'):
                    string += next_char
                else:
                    string += next_char
                self.advance()
            else:
                string += self.advance()
        if self.peek() != '"':
            self.error("Unclosed string")
        self.advance()
        return Token(TokenType.STRING, string, start_line, start_col)
    
    def read_multiline_string(self) -> Token:
        start_line = self.line
        start_col = self.column
        self.advance()
        string = ""
        while self.peek() and self.peek() != '\'':
            string += self.advance()
        if self.peek() != '\'':
            self.error("Unclosed multiline string")
        self.advance()
        return Token(TokenType.MULTILINE_STRING, string, start_line, start_col)
    
    def read_number(self) -> Token:
        start_line = self.line
        start_col = self.column
        number = ""
        if self.peek() == '-':
            number += self.advance()
            if not self.peek() or not self.peek().isdigit():
                self.error("Expected digit after '-'")
        
        has_dot = False
        has_digits_before_dot = False
        has_digits_after_dot = False
        
        while self.peek() and (self.peek().isdigit() or self.peek() == '.'):
            if self.peek() == '.':
                if has_dot:
                    break
                has_dot = True
                number += self.advance()
            else:
                number += self.advance()
                if has_dot:
                    has_digits_after_dot = True
                else:
                    has_digits_before_dot = True
        
        if not has_digits_before_dot and not has_digits_after_dot:
            self.error("Invalid number format")
        
        if has_dot:
            return Token(TokenType.FLOAT, float(number), start_line, start_col)
        else:
            return Token(TokenType.NUMBER, int(number), start_line, start_col)
    
    def read_identifier(self) -> Token:
        start_line = self.line
        start_col = self.column
        identifier = ""
        while self.peek() and (self.peek().isalnum() or self.peek() in ('_', '-')):
            identifier += self.advance()
        
        keywords = {
            "bool": TokenType.BOOL,
            "str": TokenType.STR,
            "num": TokenType.NUM,
            "fl": TokenType.FL,
            "ml": TokenType.ML,
            "class": TokenType.CLASS,
            "list": TokenType.LIST,
            "dynamic": TokenType.DYNAMIC,
        }
        boolean_values = {
            "true": True,
            "false": False,
            "yes": True,
            "no": False,
        }
        
        if identifier in keywords:
            return Token(keywords[identifier], identifier, start_line, start_col)
        elif identifier in boolean_values:
            return Token(TokenType.BOOLEAN, boolean_values[identifier], start_line, start_col)
        else:
            return Token(TokenType.IDENTIFIER, identifier, start_line, start_col)
    
    def read_identifier_with_digits(self) -> Token:
        start_line = self.line
        start_col = self.column
        identifier = ""
        while self.peek() and (self.peek().isalnum() or self.peek() in ('_', '-')):
            identifier += self.advance()
        return Token(TokenType.IDENTIFIER, identifier, start_line, start_col)
    
    def tokenize(self) -> List[Token]:
        while self.pos < len(self.text):
            self.skip_whitespace()
            if self.peek() is None:
                break
            
            if self.peek() == '[':
                self.tokens.append(self.read_comment())
                continue
            
            if self.peek() == '\n':
                newline_line = self.line
                newline_col = self.column
                self.advance()
                self.tokens.append(Token(TokenType.NEWLINE, '\n', newline_line, newline_col))
                continue
            
            if self.peek() == ':' and self.peek(1) == ':':
                start_col = self.column
                self.advance()
                self.advance()
                self.tokens.append(Token(TokenType.DOUBLE_COLON, '::', self.line, start_col))
                continue
            
            if self.peek() in '{}(),':
                start_col = self.column
                ch = self.advance()
                ttype = {
                    '{': TokenType.LBRACE,
                    '}': TokenType.RBRACE,
                    '(': TokenType.LPAREN,
                    ')': TokenType.RPAREN,
                    ',': TokenType.COMMA
                }[ch]
                self.tokens.append(Token(ttype, ch, self.line, start_col))
                continue
            
            if self.peek() == '"':
                self.tokens.append(self.read_string())
                continue
            
            if self.peek() == '\'':
                self.tokens.append(self.read_multiline_string())
                continue
            
            if self.peek() == '-' and self.peek(1) and self.peek(1).isdigit():
                self.tokens.append(self.read_number())
                continue
            
            if self.peek().isdigit():
                peek_ahead = self.pos + 1
                while peek_ahead < len(self.text) and self.text[peek_ahead].isdigit():
                    peek_ahead += 1
                if peek_ahead < len(self.text) and (self.text[peek_ahead].isalpha() or self.text[peek_ahead] == '_'):
                    self.tokens.append(self.read_identifier_with_digits())
                else:
                    self.tokens.append(self.read_number())
                continue
            
            if self.peek().isalpha() or self.peek() == '_':
                self.tokens.append(self.read_identifier())
                continue
            
            self.error(f"Unexpected character: {self.peek()}")
        
        self.tokens.append(Token(TokenType.EOF, None, self.line, self.column))
        return self.tokens


